Update to Kubernetes 1.6 and Bootkube 0.4.0

Quentin-M opened the pull request:
Good evening,

This PR updates bootkube (0.4.0), its manifests (Kubernetes 1.6.1), introduces liveness probes to CM/Scheduler and add tolerations/anti-affinity for the critical components to run next to each other but avoid running multiple times on the same node.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+coreos.0", GitCommit:"9212f77ed8c169a0afa02e58dce87913c6387b3e", GitTreeState:"clean", BuildDate:"2017-04-04T00:32:53Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

$ kubectl get nodes
NAME                                        STATUS    AGE
ip-10-0-10-234.us-west-2.compute.internal   Ready     9m
ip-10-0-19-89.us-west-2.compute.internal    Ready     9m
ip-10-0-42-19.us-west-2.compute.internal    Ready     9m
ip-10-0-54-106.us-west-2.compute.internal   Ready     9m
ip-10-0-74-57.us-west-2.compute.internal    Ready     9m

$ kubectl get pods -n=kube-system
NAME                                                               READY     STATUS    RESTARTS   AGE
heapster-961268227-wh7mh                                           2/2       Running   0          52s
kube-apiserver-984hh                                               1/1       Running   0          9m
kube-apiserver-g909c                                               1/1       Running   0          9m
kube-apiserver-r83s1                                               1/1       Running   6          9m
kube-controller-manager-1744054037-274tf                           1/1       Running   0          9m
kube-controller-manager-1744054037-g3fc2                           1/1       Running   0          9m
kube-dns-2891576090-tqf6h                                          3/3       Running   0          9m
kube-flannel-29tqb                                                 2/2       Running   2          9m
kube-flannel-5z3hm                                                 2/2       Running   3          9m
kube-flannel-k6bqk                                                 2/2       Running   3          9m
kube-flannel-qmfpq                                                 2/2       Running   3          9m
kube-flannel-tqp2w                                                 2/2       Running   2          9m
kube-proxy-0v670                                                   1/1       Running   0          7m
kube-proxy-3clvk                                                   1/1       Running   0          7m
kube-proxy-3pxlb                                                   1/1       Running   0          7m
kube-proxy-7sv92                                                   1/1       Running   0          7m
kube-proxy-n56n9                                                   1/1       Running   0          7m
kube-scheduler-2496325411-ct3jf                                    1/1       Running   0          9m
kube-scheduler-2496325411-rrw2z                                    1/1       Running   0          9m
pod-checkpointer-hmdw4                                             1/1       Running   0          3m
pod-checkpointer-hmdw4-ip-10-0-10-234.us-west-2.compute.internal   1/1       Running   0          3m
pod-checkpointer-jjcx2                                             1/1       Running   0          3m
pod-checkpointer-jjcx2-ip-10-0-42-19.us-west-2.compute.internal    1/1       Running   0          3m
pod-checkpointer-jwzq5                                             1/1       Running   0          3m
pod-checkpointer-jwzq5-ip-10-0-19-89.us-west-2.compute.internal    1/1       Running   0          3m
```

alekssaul commented on config.tf:
```diff
@@ -21,12 +21,11 @@ variable "tectonic_container_images" {
-    kubednsmasq                  = "gcr.io/google_containers/kube-dnsmasq-amd64:1.4.1"
-    dnsmasq_metrics              = "gcr.io/google_containers/dnsmasq-metrics-amd64:1.0.1"
-    exechealthz                  = "gcr.io/google_containers/exechealthz-amd64:1.2"
+    kubedns                      = "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.1"
```
Can we also bump heapster from 1.3.0-beta.0 to 1.3.0; https://github.com/kubernetes/kubernetes/blob/v1.6.1/cluster/addons/cluster-monitoring/standalone/heapster-controller.yaml

Quentin-M replied:
Thanks :)

alekssaul commented:
> Error Loading Jobs.

`extensions/v1beta1.Job` was in process of being deprecated in favor of `batch/v1` [0] . I had https://github.com/coreos-inc/tectonic-issues/issues/151 open regarding this. 

[0]https://kubernetes.io/docs/concepts/jobs/run-to-completion-finite-workloads/#extensionsv1beta1job-is-deprecated

philips commented:
cc @squat @sym3tri

Quentin-M commented:
Note: also take a look at https://github.com/kubernetes-incubator/bootkube/pull/434.

squat commented:
@Quentin-M and also kubernetes-incubator/bootkube#421 <- this includes anti-affinity for critical control plane components

Quentin-M commented:
@squat This one should be implemented already. I used git blame on templates.go - so anything that is already merged should be there.

squat commented on modules/bootkube/resources/manifests/kube-apiserver.yaml:
```diff
@@ -43,7 +49,7 @@ spec:
         - --client-ca-file=/etc/kubernetes/secrets/ca.crt
         - --authorization-mode=RBAC
-        - --runtime-config=rbac.authorization.k8s.io/v1alpha1
+        - --runtime-config=rbac.authorization.k8s.io/v1beta1
```
this should be unnecessary since all beta API groups are enabled by default

Quentin-M replied:
Smart! I didn't know that. Let me remove that line then.

squat commented on modules/bootkube/resources/bootstrap-manifests/bootstrap-apiserver.yaml:
```diff
@@ -0,0 +1,53 @@
+    - --insecure-port=8080
+    - --kubelet-client-certificate=/etc/kubernetes/secrets/apiserver.crt
+    - --kubelet-client-key=/etc/kubernetes/secrets/apiserver.key
+    - --runtime-config=api/all=true
```
We should remove this since we are not enabling any alpha API groups and do not want to unknowingly depend on them.

Quentin-M replied:
I agree. I simply copy-pasted whatever is in bootkube here without modifications but for templating purposes. Generally want to stay as close as their manifests as possible.

Quentin-M replied:
@aaronlevy Is there a specific reason this is here?

squat replied:
Yes I think so. We had a story that explicitly asks to disable allowing all alpha features by default. If a user really wants to use alpha API features then they can always modify this daemonset.

https://github.com/coreos-inc/tectonic/blob/master/docs-internal/alpha-features.md
https://www.pivotaltracker.com/story/show/137556789

aaronlevy replied:
Only exists for flexibility - but I am 100% for removing this.

aaronlevy replied:
We may need to explicitly allow TPRs though -- that is still technically an alpha feature

squat replied:
@aaronlevy TPR is v1beta1

squat replied:
https://github.com/kubernetes-incubator/bootkube/pull/440

Quentin-M replied:
Addressed.

Quentin-M commented:
/cc @amrutac. Pinged you with credentials to a Kubernetes 1.6 deployment to witness the Console breakages. Tell me if there is anything I can do to help you!

aaronlevy commented on modules/bootkube/resources/bootstrap-manifests/bootstrap-apiserver.yaml:
```diff
@@ -0,0 +1,53 @@
+    - /var/lock/api-server.lock
+    - /hyperkube
+    - apiserver
+    - --admission-control=NamespaceLifecycle,ServiceAccount
```
We should update this to the upstream recommended list: https://github.com/kubernetes-incubator/bootkube/issues/438

aaronlevy replied:
Added issue to track this generally: https://github.com/coreos/tectonic-installer/issues/249

Quentin-M replied:
Addressed.

Quentin-M commented:
https://github.com/kubernetes-incubator/bootkube/pull/434
https://github.com/kubernetes-incubator/bootkube/pull/440

s-urbaniak commented:
Setting the "do not merge" label as long as we don't have a working console.

alexsomesan commented:
@sym3tri Looks like this is causing some misbehaviour in the console? I presume we don't want to integrate it before the console has addressed those, right?

@amrutac @squat Could you please ping back here when the related console changes are in?

Quentin-M commented:
Updated Console for Kubernetes 1.6 with the early bugs I noticed on Sunday.
Updated with the Bootkube PRs https://github.com/coreos/tectonic-installer/issues/249 https://github.com/kubernetes-incubator/bootkube/pull/440 https://github.com/kubernetes-incubator/bootkube/pull/421 https://github.com/kubernetes-incubator/bootkube/pull/434
/cc @s-urbaniak @alexsomesan @sym3tri @squat @amrutac

aaronlevy commented on modules/bootkube/resources/manifests/kube-apiserver.yaml:
```diff
@@ -66,6 +65,16 @@ spec:
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
Can we just use `operator: "Exists"` here?

Quentin-M replied:
Just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go

aaronlevy commented on modules/bootkube/resources/manifests/kube-controller-manager.yaml:
```diff
@@ -27,13 +45,28 @@ spec:
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
`operator: Exists`

Quentin-M replied:
Was just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go. Fixed.

aaronlevy commented on modules/bootkube/resources/manifests/kube-dns.yaml:
```diff
@@ -140,33 +150,28 @@ spec:
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
`operator: Exists`

Quentin-M replied:
Was just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go. Fixed.

aaronlevy commented on modules/bootkube/resources/manifests/kube-flannel.yaml:
```diff
@@ -74,6 +73,12 @@ spec:
+      hostNetwork: true
+      tolerations:
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
`operator: Exists`

Quentin-M replied:
Was just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go. Fixed.

aaronlevy commented on modules/bootkube/resources/manifests/kube-proxy.yaml:
```diff
@@ -36,6 +39,16 @@ spec:
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
`operator: Exists`

Quentin-M replied:
Was just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go. Fixed.

aaronlevy commented on modules/bootkube/resources/manifests/pod-checkpointer.yaml:
```diff
@@ -0,0 +1,61 @@
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      - key: "node-role.kubernetes.io/master"
+        operator: Equal
```
`operator: Exists`

Quentin-M replied:
Was just using what you guys have in https://github.com/kubernetes-incubator/bootkube/blob/85036e7/pkg/asset/internal/templates.go. Fixed.

Quentin-M commented:
Just noticed that bootkube uses a 1.5 client, therefore the tolerations are not even applied. Have to wait.

aaronlevy commented:
I opened https://github.com/kubernetes-incubator/bootkube/issues/451 to track changing to `operator: Exists` in bootkube too

s-urbaniak commented:
Tentatively setting this as do-not-merge, as long as it is in WIP state.

Quentin-M commented:
Just bumped flannel and fixed a bunch of other stuff. Tested quite a bit by a few people on AWS - other platforms are left untested. Can we still merge and then have @squat test/follow-up later as necessary please? We need all platforms and everybody to start using it actively. We can't afford waiting until release day.

Quentin-M commented:
THANK YOU!

s-urbaniak commented:
Also tested this myself and it is looking great, thanks!!! I also quickly tested this on openstack/neutron and openstack/nova which worked fine with a minor fix on the nova-side, that I'll file as a separate PR.
